domain/charm/state/manifest.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"fmt"
+
+	"github.com/juju/juju/domain/charm"
+)
+
+// decodeManifest decodes the given manifests into a charm.Manifest.
+// It should respect the order of the bases in the manifest. Although the
+// order of architectures within a base is not guaranteed.
+func decodeManifest(manifests []charmManifest) (charm.Manifest, error) {
```

Comment:
Is it worth making these methods on a custom type of `[]charmManifest`? That's been the convention in other domains.